GetStream/stream-chat-android
# [4422] Improve file uploader

### 🎯 Goal

Closes https://github.com/GetStream/stream-chat-android/issues/4422

When uploading images using custom CDN, clients should be able to return both original image and image thumb.

### 🧪 Testing

1. Apply the patch below
2. Upload an image
3. Ensure that  low-res image is displayed in the message list

<details>

<summary>Custom FileUploader</summary>

```
Index: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/ChatInitializer.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/ChatInitializer.kt b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/ChatInitializer.kt
--- a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/ChatInitializer.kt	(revision e6b8275747fbebaa6715b6cf2bf5e1d5b09135b6)
+++ b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/ChatInitializer.kt	(date 1668457400945)
@@ -20,8 +20,13 @@
 import com.google.firebase.FirebaseApp
 import io.getstream.chat.android.client.ChatClient
 import io.getstream.chat.android.client.logger.ChatLogLevel
+import io.getstream.chat.android.client.models.UploadedFile
+import io.getstream.chat.android.client.models.UploadedImage
 import io.getstream.chat.android.client.notifications.handler.NotificationConfig
 import io.getstream.chat.android.client.notifications.handler.NotificationHandlerFactory
+import io.getstream.chat.android.client.uploader.FileUploader
+import io.getstream.chat.android.client.utils.ProgressCallback
+import io.getstream.chat.android.client.utils.Result
 import io.getstream.chat.android.markdown.MarkdownTextTransformer
 import io.getstream.chat.android.offline.plugin.configuration.Config
 import io.getstream.chat.android.offline.plugin.factory.StreamOfflinePluginFactory
@@ -31,6 +36,7 @@
 import io.getstream.chat.android.ui.ChatUI
 import io.getstream.chat.ui.sample.BuildConfig
 import io.getstream.chat.ui.sample.feature.HostActivity
+import java.io.File
 
 class ChatInitializer(private val context: Context) {
 
@@ -77,9 +83,59 @@
             .logLevel(logLevel)
             .withPlugin(offlinePlugin)
             .debugRequests(true)
+            .fileUploader(DummyFileUploader())
             .build()
 
         // Using markdown as text transformer
         ChatUI.messageTextTransformer = MarkdownTextTransformer(context)
     }
 }
+
+private class DummyFileUploader : FileUploader {
+
+    override fun sendImage(
+        channelType: String,
+        channelId: String,
+        userId: String,
+        file: File,
+        callback: ProgressCallback,
+    ): Result<UploadedImage> {
+        return Result.success(
+            UploadedImage(
+                file = "https://picsum.photos/id/237/200/300",
+                thumbUrl = "https://picsum.photos/id/237/40/60"
+            )
+        )
+    }
+
+    override fun sendImage(channelType: String, channelId: String, userId: String, file: File): Result<UploadedImage> {
+        return Result.success(
+            UploadedImage(
+                file = "https://picsum.photos/id/237/200/300",
+                thumbUrl = "https://picsum.photos/id/237/40/60"
+            )
+        )
+    }
+
+    override fun sendFile(
+        channelType: String,
+        channelId: String,
+        userId: String,
+        file: File,
+        callback: ProgressCallback,
+    ): Result<UploadedFile> {
+        TODO("Not yet implemented")
+    }
+
+    override fun sendFile(channelType: String, channelId: String, userId: String, file: File): Result<UploadedFile> {
+        TODO("Not yet implemented")
+    }
+
+    override fun deleteFile(channelType: String, channelId: String, userId: String, url: String): Result<Unit> {
+        TODO("Not yet implemented")
+    }
+
+    override fun deleteImage(channelType: String, channelId: String, userId: String, url: String): Result<Unit> {
+        TODO("Not yet implemented")
+    }
+}

```

</details>

https://user-images.githubusercontent.com/9600921/201759171-3e22cb0c-bfeb-4513-9c3c-db35ab79e4b5.mp4

### ☑️Contributor Checklist

#### General

#### Code & documentation

### ☑️Reviewer Checklist
